File: app/src/test/java/com/mohsenoid/rickandmorty/data/RepositoryQueryCharacterDetailsTest.java
package com.mohsenoid.rickandmorty.data;

import org.json.JSONException;
import org.junit.Test;

import java.io.IOException;

import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

public class RepositoryQueryCharacterDetailsTest extends RepositoryTest {

    @Test
    public void testIfQueryCharacterDetailsCallsApiClientMethodWhenIsOnline() throws IOException, JSONException {
        // GIVEN
        stubConfigProviderIsOnline(true);
        int page = 1;

        // WHEN
        repository.queryCharacterDetails(page, null);

        // THEN
        verify(db, times(1)).queryCharacter(page);
        verify(networkClient, times(1)).getCharacterDetails(page);
    }

    @Test
    public void testIfQueryCharacterDetailsCallsApiClientMethodWhenIsOffline() throws IOException, JSONException {
        // GIVEN
        stubConfigProviderIsOnline(false);
        int page = 1;

        // WHEN
        repository.queryCharacterDetails(page, null);

        // THEN
        verify(db, times(1)).queryCharacter(page);
        verify(networkClient, times(0)).getCharacterDetails(page);
    }
}

File: app/src/test/kotlin/com/mohsenoid/rickandmorty/data/RepositoryQueryCharacterDetailsTest.kt
package com.mohsenoid.rickandmorty.data

import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.verify
import kotlinx.coroutines.runBlocking
import org.junit.Test

class RepositoryQueryCharacterDetailsTest : RepositoryTest() {

    @Test
    fun `test if queryCharacterDetails calls networkClient when isOnline`() {
        runBlocking {
            // GIVEN
            stubConfigProviderIsOnline(true)
            val page = 1

            // WHEN
            repository.queryCharacterDetails(page, null)

            // THEN
            verify(characterDao, times(1)).queryCharacter(page)
            verify(networkClient, times(1)).getCharacterDetails(page)
        }
    }

    @Test
    fun `test if queryCharacterDetails calls db only when isOffline`() {
        runBlocking {
            // GIVEN
            stubConfigProviderIsOnline(false)
            val page = 1

            // WHEN
            repository.queryCharacterDetails(page, null)

            // THEN
            verify(characterDao, times(1)).queryCharacter(page)
            verify(networkClient, times(0)).getCharacterDetails(page)
        }
    }
}
